

<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" />
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  
  <title>Pomoxis - bioinformatics tools for nanopore research &mdash; Pomoxis 0.3.6 documentation</title>
  

  
  <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  <link rel="stylesheet" href="_static/pygments.css" type="text/css" />

  
  

  
  

  

  
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
    
      <script type="text/javascript" id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    
    <script type="text/javascript" src="_static/js/theme.js"></script>

    
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Pomoxis Programs" href="programs.html" /> 
</head>

<body class="wy-body-for-nav">

   
  <div class="wy-grid-for-nav">
    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
          

          
            <a href="#" class="icon icon-home"> Pomoxis
          

          
          </a>

          
            
            
              <div class="version">
                0.3.6
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        
        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Pomoxis Programs</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="pomoxis.html">pomoxis package</a></li>
</ul>

            
          
        </div>
        
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="#">Pomoxis</a>
        
      </nav>


      <div class="wy-nav-content">
        
        <div class="rst-content">
        
          

















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="#" class="icon icon-home"></a> &raquo;</li>
        
      <li>Pomoxis - bioinformatics tools for nanopore research</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="_sources/index.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="pomoxis-bioinformatics-tools-for-nanopore-research">
<h1>Pomoxis - bioinformatics tools for nanopore research<a class="headerlink" href="#pomoxis-bioinformatics-tools-for-nanopore-research" title="Permalink to this headline">¶</a></h1>
<p>© 2018 Oxford Nanopore Technologies Ltd.</p>
<p>Pomoxis comprises a set of basic bioinformatic tools tailored to nanopore
sequencing. Notably tools are included for generating and analysing draft
assemblies. Many of these tools are used by the research data analysis
group at Oxford Nanopore Technologies.</p>
<p>See <a class="reference internal" href="programs.html"><span class="doc">Pomoxis Programs</span></a> for common simple tasks.</p>
<div class="section" id="features">
<h2>Features<a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h2>
<blockquote>
<div><ul class="simple">
<li><p>Wraps third party tools with known good default parameters
and methods of use.</p></li>
<li><p>Creates an isolated environment with all third-party tools.</p></li>
<li><p>Can be installed with conda.</p></li>
<li><p>Streamlines common short analysis chains.</p></li>
<li><p>Integrates into <a class="reference external" href="https://github.com/nanoporetech/katuali">katuali</a>
for performing more complex analysis pipelines.</p></li>
<li><p>Open source (Mozilla Public License 2.0).</p></li>
</ul>
</div></blockquote>
<div class="admonition-research-release admonition">
<p class="admonition-title">Research Release</p>
<p>Research releases are provided as technology demonstrators to provide early
access to features or stimulate Community development of tools. Support for
this software will be minimal and is only provided directly by the developers.
Feature requests, improvements, and discussions are welcome and can be
implemented by forking and pull requests. However much as we would
like to rectify every issue and piece of feedback users may have, the
developers may have limited resource for support of this software. Research
releases may be unstable and subject to rapid iteration by Oxford Nanopore
Technologies.</p>
</div>
</div>
<div class="section" id="compatibility">
<h2>Compatibility<a class="headerlink" href="#compatibility" title="Permalink to this headline">¶</a></h2>
<p>Pomoxis is developed on Ubuntu 16.04, other recent linuxes should be
equally compatible (see build notes below). Pomoxis is known to work on
at least some MacOS High Sierra configurations, though some components,
notably scrappy, are known to not work on some MacOS configurations
(combinations of OS and xcode versions).</p>
</div>
<div class="section" id="installation">
<h2>Installation<a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h2>
<p>Much of pomoxis’s functionality is dependent on thirty party tools. These
can be provided by the user, or can be installed with the help of the
provided <cite>Makefile</cite></p>
<p><strong>Installation with conda</strong></p>
<p>Pomoxis is available on <a class="reference external" href="https://bioconda.github.io/recipes/pomoxis/">bioconda</a>
and so can be most easily installed with:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>conda install pomoxis
</pre></div>
</div>
<p><strong>Installation with pip</strong></p>
<p>Pomoxis can be installed using the python package manager, pip:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install git+https://github.com/rrwick/Porechop
pip install pomoxis
</pre></div>
</div>
<p>We recommend using medaka within a virtual environment, viz.:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>virtualenv pomoxis --python<span class="o">=</span>python3 --prompt <span class="s2">&quot;(pomoxis) &quot;</span>
. pomoxis/bin/activate
pip install git+https://github.com/rrwick/Porechop
pip install pomoxis
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Using this method requires the user to provide several binaries:</p>
<p><a class="reference external" href="https://github.com/lh3/minimap2">minimap2</a>,
<a class="reference external" href="https://github.com/lh3/miniasm">miniasm</a>,
<a class="reference external" href="https://github.com/samtools/samtools">samtools</a>,
<a class="reference external" href="https://github.com/isovic/racon">racon</a>,
<a class="reference external" href="https://github.com/samtools/samtools">samtools</a>,
<a class="reference external" href="https://github.com/samtools/bcftools/">bcftools</a>, and
<a class="reference external" href="https://github.com/shenwei356/seqkiti">seqkit</a>,</p>
<p>and place these within the <cite>PATH</cite>.</p>
</div>
<p><strong>Installation from source</strong></p>
<p>With this method pomoxis will install itself into a an isolated virtual
environment. The installation will fetch, compile, and install all direct
dependencies into the environment. Use this method if you do not wish to
use conda, but will not be providing the third-party binaries.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Before installing pomoxis is may be required to install some prerequisite
packages, best installed by a package manager. On Ubuntu these are:</p>
<p>gcc-4.9 g++-4.9 zlib1g-dev libncurses5-dev python3-all-dev libhdf5-dev
libatlas-base-dev libopenblas-base libopenblas-dev libbz2-dev liblzma-dev
libffi-dev make python-virtualenv cmake wget bzip2</p>
</div>
<p>To setup the environment run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>git clone --recursive https://github.com/nanoporetech/pomoxis
<span class="nb">cd</span> pomoxis
make install
. ./venv/bin/activate
</pre></div>
</div>
<p>The installation of porechop (<a class="reference external" href="https://github.com/rrwick/Porechop">https://github.com/rrwick/Porechop</a>)
requires a newer compiler than is a available on some systems. It may therefore
be necessary to install a newer compiler and set environment variables before
the <cite>make install</cite> step:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="c1"># For porechop to be compiled on older systems set these, e.g.:</span>
<span class="nb">export</span> <span class="nv">CXX</span><span class="o">=</span><span class="s2">&quot;g++-4.9&quot;</span> <span class="nv">CC</span><span class="o">=</span><span class="s2">&quot;gcc-4.9&quot;</span>
</pre></div>
</div>
<p>Note also that racon requires at least <cite>gcc&gt;=4.8.5</cite> to
[compile smoothly](<a class="reference external" href="https://github.com/isovic/racon/issues/57">https://github.com/isovic/racon/issues/57</a>).</p>
<p><strong>Installation without compiling third-party binaries</strong></p>
<p>Running the above within a pre-exisiting (virtual) environnment may well fail;
advanced may wish to simply run</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>python setup.py install
</pre></div>
</div>
<p>in the standard manner after compiling the third party programs listed below
and ensuring they are present on the <cite>PATH</cite>. The <cite>setup.py</cite> script can copy
the binaries into the python interpreter path if they are placed within a
directory named <cite>bincache</cite> alongside <cite>setup.py</cite>. To make use of this facility
run:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>pip install -r requirements.txt
<span class="nv">POMO_BINARIES</span><span class="o">=</span><span class="m">1</span> python setup.py install
</pre></div>
</div>
</div>
<div class="section" id="third-party-binaries">
<h2>Third party binaries<a class="headerlink" href="#third-party-binaries" title="Permalink to this headline">¶</a></h2>
<p>The distribution bundles some common bioinformatics tools:</p>
<blockquote>
<div><ul class="simple">
<li><p>miniasm</p></li>
<li><p>minimap2</p></li>
<li><p>racon</p></li>
<li><p>samtools</p></li>
<li><p>bcftools</p></li>
<li><p>seqkit</p></li>
<li><p>porechop</p></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="contents">
<h2>Contents<a class="headerlink" href="#contents" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="programs.html">Pomoxis Programs</a><ul>
<li class="toctree-l2"><a class="reference internal" href="programs.html#assess-assembly">assess_assembly</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#assess-homopolymers">assess_homopolymers</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#catalogue-errors">catalogue_errors</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#common-errors-from-bam">common_errors_from_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#coverage-from-bam">coverage_from_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#coverage-from-fastx">coverage_from_fastx</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#fast-convert">fast_convert</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#find-indels">find_indels</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#intersect-assembly-errors">intersect_assembly_errors</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#long-fastx">long_fastx</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#mini-align">mini_align</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#mini-assemble">mini_assemble</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#pomoxis-path">pomoxis_path</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#qscores-from-summary">qscores_from_summary</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#ref-seqs-from-bam">ref_seqs_from_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#reverse-bed">reverse_bed</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#split-fastx">split_fastx</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#stats-from-bam">stats_from_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#subsample-bam">subsample_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#summary-from-stats">summary_from_stats</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#tag-bam">tag_bam</a></li>
<li class="toctree-l2"><a class="reference internal" href="programs.html#trim-alignments">trim_alignments</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="full-api-reference">
<h2>Full API reference<a class="headerlink" href="#full-api-reference" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="pomoxis.html">pomoxis package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.assess_homopolymers">pomoxis.assess_homopolymers module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.bio">pomoxis.bio module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.catalogue_errors">pomoxis.catalogue_errors module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.common_errors_from_bam">pomoxis.common_errors_from_bam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.coverage_from_bam">pomoxis.coverage_from_bam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.find_indels">pomoxis.find_indels module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.qscores_from_summary">pomoxis.qscores_from_summary module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.ref_seqs_from_bam">pomoxis.ref_seqs_from_bam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.stats_from_bam">pomoxis.stats_from_bam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.subsample_bam">pomoxis.subsample_bam module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.summary_from_stats">pomoxis.summary_from_stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.trim_alignments">pomoxis.trim_alignments module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis.util">pomoxis.util module</a></li>
<li class="toctree-l2"><a class="reference internal" href="pomoxis.html#module-pomoxis">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="indices-and-tables">
<h2>Indices and tables<a class="headerlink" href="#indices-and-tables" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="genindex.html"><span class="std std-ref">Index</span></a></p></li>
<li><p><a class="reference internal" href="py-modindex.html"><span class="std std-ref">Module Index</span></a></p></li>
<li><p><a class="reference internal" href="search.html"><span class="std std-ref">Search Page</span></a></p></li>
</ul>
</div>
</div>


           </div>
           
          </div>
          <footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
        <a href="programs.html" class="btn btn-neutral float-right" title="Pomoxis Programs" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>
        &#169; Copyright 2017, Oxford Nanopore Technologies.

    </p>
  </div>
    
    
    
    Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    
    provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>
        </div>
      </div>

    </section>

  </div>
  

  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script>

  
  
    
   

</body>
</html>